File: kits/cpp/src/lux/factory.cpp
#include "lux/factory.hpp"

#include <numeric>

#include "lux/config.hpp"
#include "lux/observation.hpp"

namespace lux {
    int64_t Factory::buildMetalCost(const Observation &obs,

                                    const std::string &unitType) const {
        return obs.config.ROBOTS[unitType].METAL_COST;
    }

    int64_t Factory::buildPowerCost(const Observation &obs,
                                    const std::string &unitType) const {
        auto weather = obs.getCurrentWeather();
        return std::ceil(obs.config.ROBOTS[unitType].POWER_COST * weather.POWER_CONSUMPTION);
    }

    bool Factory::canBuild(const Observation &obs, const std::string &unitType) const {
        return power >= buildPowerCost(obs, unitType) && cargo.metal >= buildMetalCost(obs, unitType);
    }

    int64_t Factory::buildHeavyMetalCost(const Observation &obs) const { return buildMetalCost(obs, "HEAVY"); }

    int64_t Factory::buildHeavyPowerCost(const Observation &obs) const { return buildPowerCost(obs, "HEAVY"); }

    bool Factory::canBuildHeavy(const Observation &obs) const { return canBuild(obs, "HEAVY"); }

    FactoryAction Factory::buildHeavy(const Observation &obs) const {
        UNUSED(obs);
        LUX_ASSERT(canBuildHeavy(obs), "factory cannot build HEAVY, make sure to check beforehand with canBuildHeavy");
        return FactoryAction::BuildHeavy();
    }

    int64_t Factory::buildLightMetalCost(const Observation &obs) const { return buildMetalCost(obs, "LIGHT"); }

    int64_t Factory::buildLightPowerCost(const Observation &obs) const { return buildPowerCost(obs, "LIGHT"); }

    bool Factory::canBuildLight(const Observation &obs) const { return canBuild(obs, "LIGHT"); }

    FactoryAction Factory::buildLight(const Observation &obs) const {
        UNUSED(obs);
        LUX_ASSERT(canBuildLight(obs), "factory cannot build LIGHT, make sure to check beforehand with canBuildLight");
        return FactoryAction::BuildLight();
    }

    int64_t Factory::waterCost(const Observation &obs) const {
        auto countMatchingStrains = [&](int64_t sum, auto row) {
            return sum + std::count(row.begin(), row.end(), strain_id);
        };
        int64_t sum =
            std::accumulate(obs.board.lichen_strains.begin(), obs.board.lichen_strains.end(), 0, countMatchingStrains);
        return std::ceil(sum / obs.config.LICHEN_WATERING_COST_FACTOR) + 1;
    }

    bool Factory::canWater(const Observation &obs) const { return cargo.water >= waterCost(obs); }

    FactoryAction Factory::water(const Observation &obs) const {
        UNUSED(obs);
        LUX_ASSERT(canWater(obs), "factory cannot water, make sure to check beforehand with canWater");
        return FactoryAction::Water();
    }
}  // namespace lux

File: kits/rust/lux/src/factory.rs
//! Handles serializing, deserializing and interacting with factory units

use crate::cargo::Cargo;
use crate::config::Config;
use crate::state::State;
use crate::Pos;
use serde::{Deserialize, Serialize};
use std::ops::Range;

/// A representation of a Factory
///
/// A factory is a 3x3 unit that can build units, and grow lichen
///
/// For more information see the [spec](https://www.lux-ai.org/specs-s2#factories)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Factory {
    /// Id discriminating the team Factory belongs to
    ///
    /// valid values are 0 or 1
    pub team_id: u64,

    /// Unique identfier for the factory
    ///
    /// Will be of the form "factory_<unique_index>"
    pub unit_id: String,

    /// Current factory power
    pub power: u64,

    /// Position of c
    pub pos: Pos,

    /// Current materials stored in the factory
    pub cargo: Cargo,

    /// Unique identifier for the lichen strain this factory grows
    pub strain_id: u64,
}

impl Factory {
    /// Gets 3x3 range as two range iterators
    #[inline]
    pub fn occupied_range(&self) -> (Range<i64>, Range<i64>) {
        (
            (self.pos.0 - 1..self.pos.0 + 2),
            (self.pos.1 - 1..self.pos.1 + 2),
        )
    }

    /// Validates if the factory has enough power and metal to build a heavy robot
    #[inline]
    pub fn can_build_heavy(&self, cfg: &Config) -> bool {
        let metal_cost = cfg.robots.heavy.metal_cost;
        let power_cost = cfg.robots.heavy.power_cost;
        self.power >= power_cost && self.cargo.metal >= metal_cost
    }

    /// Validates if the factory has enough power and metal to build a light robot
    #[inline]
    pub fn can_build_light(&self, cfg: &Config) -> bool {
        let metal_cost = cfg.robots.light.metal_cost;
        let power_cost = cfg.robots.light.power_cost;
        self.power >= power_cost && self.cargo.metal >= metal_cost
    }

    /// Gets power cost to water lichen
    ///
    /// # Note
    ///
    /// This associated function is an overestimate of the cost as the
    /// true calculation requires the growth set of tiles rather than the
    /// existing tiles. This matches the python implementation and as such
    /// there are no plans to calculate the real cost from this function
    #[inline]
    pub fn water_cost(&self, state: &State) -> u64 {
        let tile_count = state
            .board
            .tiles
            .iter()
            .filter(|tile| {
                if let Some(lichen) = &tile.lichen {
                    lichen.strain == self.strain_id
                } else {
                    false
                }
            })
            .count() as f64;
        f64::ceil(tile_count / state.env_cfg.lichen_watering_cost_factor) as u64
    }
}
